CORGI-805 add back cdn relations

jasinner opened the pull request:
This essentially reverts the changes made for CORGI-727. I kept some of the refactoring, and cleaned up the code a little bit. This should get coverage for RHEL Z-Streams back up to parity with Deptopia.

To execute the code after merging, without waiting for the weekly task, manually run `setup_pulp_relations` task, and then `loadpulpdata` management command for the desire stream(s).

jasinner committed:
revert CORGI-727/b50aaa2

jasinner commented:
@RedHatProductSecurity/corgi-devs please review

juspence commented on corgi/tasks/yum.py:
```diff
@@ -9,14 +9,15 @@
     create_relations,
     get_last_success_for_task,
 )
+from corgi.tasks.pulp import fetch_unprocessed_relations
```
You can leave off this change - it was actually a typo where I imported fetch_unprocessed_relations from the corgi.tasks.pulp module instead of from the corgi.tasks.brew module.

It worked because pulp also imported fetch_unprocessed_relations from brew, but then it broke when we removed that import from the pulp module. Let's continue importing the module from where it really is defined, so we avoid confusion and possible breakage in the future.

jasinner replied:
ok

juspence commented on corgi/tasks/pulp.py:
```diff
@@ -1,14 +1,89 @@
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_setup_pulp_rpm_relations(channel, variant):
```
It's not related to this change, but maybe we could add type hints here while we're making changes. By default, mypy will ignore and not check any functions without type hits, so even something as simple as adding a `-> None` return type hint improves our mypy coverage.

jasinner replied:
The create_relation function is a bit tricky to figure out the correct types, so I'm going to skip that for now.

juspence commented on corgi/tasks/pulp.py:
```diff
@@ -1,14 +1,89 @@
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_setup_pulp_module_relations(channel, variant):
```
I'm not sure if we know the types or if mypy will start throwing some complicated error. If it's a problem we can leave off type hints for now.

jasinner replied:
leaving off for now.

juspence commented on corgi/tasks/management/commands/loadpulpdata.py:
```diff
@@ -0,0 +1,58 @@
+                self.get_builds_by_cdn_repo(stream_name=stream_name, force_process=options["force"])
+        elif options["all"]:
+            self.stdout.write(self.style.NOTICE("Fetching all unprocessed pulp relations"))
+            fetch_unprocessed_cdn_relations(force_process=options["force"])
```
Is it expected that the command runs this Celery task inline? Should we use .delay() to schedule it for later instead?

It's fine to leave for now, I just ask since this is the only time I know I'll be looking at this code. I don't want to accidentally add back a bug (?) we might have had without noticing before.

jasinner replied:
good idea. Adding .delay() call to this, and the fetch_modular_builds call below.

juspence reviewed (approved):
This code looks...familiar :grin:  I don't fully remember what I removed, but this looks like enough of it, and I assume you're looking at the diff so haven't missed anything.

I also don't remember what the side-effects were when we removed this the first time. Do we need to change something about the taxonomy-saving code to handle the CDN_REPO relations, or do we need to write some migration for something?

Sorry if that's really vague, I just remember the removal being more complicated so I thought adding it back wouldn't be this simple. I guess we can merge this, run the tasks, and see if all the relations get created / components get linked to products correctly now. Thanks for fixing this!

jasinner commented:
> This code looks...familiar 😁 I don't fully remember what I removed, but this looks like enough of it, and I assume you're looking at the diff so haven't missed anything.
> 
> I also don't remember what the side-effects were when we removed this the first time. Do we need to change something about the taxonomy-saving code to handle the CDN_REPO relations, or do we need to write some migration for something?
> 
> Sorry if that's really vague, I just remember the removal being more complicated so I thought adding it back wouldn't be this simple. I guess we can merge this, run the tasks, and see if all the relations get created / components get linked to products correctly now. Thanks for fixing this!

There was a related change CORGI-778 which adjusted the save_product_taxonomy function. That's probably the only change we really needed, not this one.
